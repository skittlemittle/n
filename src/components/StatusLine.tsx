import * as React from "react";
import styled, { ThemeContext } from "styled-components";
import { STATUSLINE_COLORS } from "../themeConstants";
import StatusSection from "./StatusSection";

const Line = styled.div.attrs((props: { bg_color: string }) => ({
  bg_color: props.bg_color,
}))`
  display: flex;
  flex-direction: row;
  background-color: ${(props) => props.theme.colors[props.bg_color]};
`;

const Right = styled.div`
  display: flex;
  margin-left: auto;
  flex-direction: row-reverse;
`;

function MakeSection(
  content: string,
  index: number,
  colors: sectionColors,
  facing: "left" | "right"
) {
  const themeContext = React.useContext(ThemeContext);
  return (
    <StatusSection
      dir={facing}
      textColor={themeContext.colors[colors.fg]}
      color={themeContext.colors[colors.bg]}
      index={index}
      key={index}
    >
      {content}
    </StatusSection>
  );
}

interface props {
  stats: DocStats;
}

/** statusline */
const StatusLine = ({ stats }: props) => {
  const colors = STATUSLINE_COLORS[stats.mode].sections;
  const bg = STATUSLINE_COLORS[stats.mode].background;

  return (
    <Line bg_color={bg}>
      {stats.left.map((stat, i) => MakeSection(stat, i, colors[i], "right"))}
      <Right>
        {stats.right.map((stat, i) => MakeSection(stat, i, colors[i], "left"))}
      </Right>
    </Line>
  );
};

export default StatusLine;

/** left / right: string arrays, each string is one section */
interface DocStats {
  /** all lowercase, valid mode name */
  mode: StatusMode;
  left: string[];
  right: string[];
}

type StatusMode = "insert" | "visual" | "rendered" | "normal";

interface sectionColors {
  bg: string;
  fg: string;
}
