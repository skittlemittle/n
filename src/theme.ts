import { COLORS_L } from "./themeConstants";

export const AppTheme = {
  colors: COLORS_L,
};
