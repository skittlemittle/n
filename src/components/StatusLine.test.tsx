import renderer from "react-test-renderer";
import { ThemeProvider } from "styled-components";
import { AppTheme } from "../theme";
import StatusLine from "./StatusLine";

it("insert colors", () => {
  const tree = renderer
    .create(
      <ThemeProvider theme={AppTheme}>
        <StatusLine
          stats={{
            mode: "insert",
            left: ["INSERT", "some file", "8?"],
            right: ["3% : 3/300", "hah", "hheh"],
          }}
        />
      </ThemeProvider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it("rendered colors", () => {
  const tree = renderer
    .create(
      <ThemeProvider theme={AppTheme}>
        <StatusLine
          stats={{
            mode: "rendered",
            left: ["RENDERED"],
            right: ["3% : 3/300"],
          }}
        />
      </ThemeProvider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it("visual colors", () => {
  const tree = renderer
    .create(
      <ThemeProvider theme={AppTheme}>
        <StatusLine
          stats={{
            mode: "visual",
            left: ["V_BLOCK", "owo", "3"],
            right: ["3% : 3/300", "ae", "r"],
          }}
        />
      </ThemeProvider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it("normal colors", () => {
  const tree = renderer
    .create(
      <ThemeProvider theme={AppTheme}>
        <StatusLine
          stats={{
            mode: "normal",
            left: ["NORMAL", "hm", "3"],
            right: ["3% : 3/300", "ae", "r"],
          }}
        />
      </ThemeProvider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
