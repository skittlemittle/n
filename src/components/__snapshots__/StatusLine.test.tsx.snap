// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`insert colors 1`] = `
<div
  className="sc-hKwDye bDPwjS"
>
  <div
    style={
      Object {
        "display": "flex",
        "zIndex": 100,
      }
    }
  >
    <div
      className="sc-gsDKAQ sc-dkPtRN ktNGwk hjGvrY"
      dir="left"
    >
      INSERT
    </div>
    <div
      className="sc-bdvvtL cXyHrn"
      color="#076678"
      dir="left"
    />
  </div>
  <div
    style={
      Object {
        "display": "flex",
        "zIndex": 99,
      }
    }
  >
    <div
      className="sc-gsDKAQ sc-dkPtRN ktNGwk drgQRY"
      dir="left"
    >
      some file
    </div>
    <div
      className="sc-bdvvtL dsnwVd"
      color="#d5c4a1"
      dir="left"
    />
  </div>
  <div
    style={
      Object {
        "display": "flex",
        "zIndex": 98,
      }
    }
  >
    <div
      className="sc-gsDKAQ sc-dkPtRN ktNGwk iXDobc"
      dir="left"
    >
      8?
    </div>
    <div
      className="sc-bdvvtL dsnwVd"
      color="#d5c4a1"
      dir="left"
    />
  </div>
  <div
    className="sc-eCImPb iaLjNI"
  >
    <div
      style={
        Object {
          "display": "flex",
          "zIndex": 100,
        }
      }
    >
      <div
        className="sc-bdvvtL efovQs"
        color="#076678"
        dir="right"
      />
      <div
        className="sc-gsDKAQ sc-dkPtRN ktNGwk edMYak"
        dir="right"
      >
        3% : 3/300
      </div>
    </div>
    <div
      style={
        Object {
          "display": "flex",
          "zIndex": 99,
        }
      }
    >
      <div
        className="sc-bdvvtL xxeWu"
        color="#d5c4a1"
        dir="right"
      />
      <div
        className="sc-gsDKAQ sc-dkPtRN ktNGwk hCHYao"
        dir="right"
      >
        hah
      </div>
    </div>
    <div
      style={
        Object {
          "display": "flex",
          "zIndex": 98,
        }
      }
    >
      <div
        className="sc-bdvvtL xxeWu"
        color="#d5c4a1"
        dir="right"
      />
      <div
        className="sc-gsDKAQ sc-dkPtRN ktNGwk fkdtWg"
        dir="right"
      >
        hheh
      </div>
    </div>
  </div>
</div>
`;

exports[`normal colors 1`] = `
<div
  className="sc-hKwDye hlksiN"
>
  <div
    style={
      Object {
        "display": "flex",
        "zIndex": 100,
      }
    }
  >
    <div
      className="sc-gsDKAQ sc-dkPtRN ktNGwk cSLGga"
      dir="left"
    >
      NORMAL
    </div>
    <div
      className="sc-bdvvtL bYKFuF"
      color="#7c6f64"
      dir="left"
    />
  </div>
  <div
    style={
      Object {
        "display": "flex",
        "zIndex": 99,
      }
    }
  >
    <div
      className="sc-gsDKAQ sc-dkPtRN ktNGwk iXDobc"
      dir="left"
    >
      hm
    </div>
    <div
      className="sc-bdvvtL dsnwVd"
      color="#d5c4a1"
      dir="left"
    />
  </div>
  <div
    style={
      Object {
        "display": "flex",
        "zIndex": 98,
      }
    }
  >
    <div
      className="sc-gsDKAQ sc-dkPtRN ktNGwk cviyfT"
      dir="left"
    >
      3
    </div>
    <div
      className="sc-bdvvtL eCTkIa"
      color="#ebdbb2"
      dir="left"
    />
  </div>
  <div
    className="sc-eCImPb iaLjNI"
  >
    <div
      style={
        Object {
          "display": "flex",
          "zIndex": 100,
        }
      }
    >
      <div
        className="sc-bdvvtL jMyhKK"
        color="#7c6f64"
        dir="right"
      />
      <div
        className="sc-gsDKAQ sc-dkPtRN ktNGwk gNWHXG"
        dir="right"
      >
        3% : 3/300
      </div>
    </div>
    <div
      style={
        Object {
          "display": "flex",
          "zIndex": 99,
        }
      }
    >
      <div
        className="sc-bdvvtL xxeWu"
        color="#d5c4a1"
        dir="right"
      />
      <div
        className="sc-gsDKAQ sc-dkPtRN ktNGwk fkdtWg"
        dir="right"
      >
        ae
      </div>
    </div>
    <div
      style={
        Object {
          "display": "flex",
          "zIndex": 98,
        }
      }
    >
      <div
        className="sc-bdvvtL gICKSd"
        color="#ebdbb2"
        dir="right"
      />
      <div
        className="sc-gsDKAQ sc-dkPtRN ktNGwk dPiJWH"
        dir="right"
      >
        r
      </div>
    </div>
  </div>
</div>
`;

exports[`rendered colors 1`] = `
<div
  className="sc-hKwDye bDPwjS"
>
  <div
    style={
      Object {
        "display": "flex",
        "zIndex": 100,
      }
    }
  >
    <div
      className="sc-gsDKAQ sc-dkPtRN ktNGwk cpVntN"
      dir="left"
    >
      RENDERED
    </div>
    <div
      className="sc-bdvvtL ddHHXW"
      color="#8f3f71"
      dir="left"
    />
  </div>
  <div
    className="sc-eCImPb iaLjNI"
  >
    <div
      style={
        Object {
          "display": "flex",
          "zIndex": 100,
        }
      }
    >
      <div
        className="sc-bdvvtL IdxZ"
        color="#8f3f71"
        dir="right"
      />
      <div
        className="sc-gsDKAQ sc-dkPtRN ktNGwk ibfvOF"
        dir="right"
      >
        3% : 3/300
      </div>
    </div>
  </div>
</div>
`;

exports[`visual colors 1`] = `
<div
  className="sc-hKwDye loAIBN"
>
  <div
    style={
      Object {
        "display": "flex",
        "zIndex": 100,
      }
    }
  >
    <div
      className="sc-gsDKAQ sc-dkPtRN ktNGwk iBcGEW"
      dir="left"
    >
      V_BLOCK
    </div>
    <div
      className="sc-bdvvtL lmFPxZ"
      color="#af3a03"
      dir="left"
    />
  </div>
  <div
    style={
      Object {
        "display": "flex",
        "zIndex": 99,
      }
    }
  >
    <div
      className="sc-gsDKAQ sc-dkPtRN ktNGwk iXDobc"
      dir="left"
    >
      owo
    </div>
    <div
      className="sc-bdvvtL dsnwVd"
      color="#d5c4a1"
      dir="left"
    />
  </div>
  <div
    style={
      Object {
        "display": "flex",
        "zIndex": 98,
      }
    }
  >
    <div
      className="sc-gsDKAQ sc-dkPtRN ktNGwk eSLkxa"
      dir="left"
    >
      3
    </div>
    <div
      className="sc-bdvvtL jvswJW"
      color="#a89984"
      dir="left"
    />
  </div>
  <div
    className="sc-eCImPb iaLjNI"
  >
    <div
      style={
        Object {
          "display": "flex",
          "zIndex": 100,
        }
      }
    >
      <div
        className="sc-bdvvtL fVvAoC"
        color="#af3a03"
        dir="right"
      />
      <div
        className="sc-gsDKAQ sc-dkPtRN ktNGwk gCEJlO"
        dir="right"
      >
        3% : 3/300
      </div>
    </div>
    <div
      style={
        Object {
          "display": "flex",
          "zIndex": 99,
        }
      }
    >
      <div
        className="sc-bdvvtL xxeWu"
        color="#d5c4a1"
        dir="right"
      />
      <div
        className="sc-gsDKAQ sc-dkPtRN ktNGwk fkdtWg"
        dir="right"
      >
        ae
      </div>
    </div>
    <div
      style={
        Object {
          "display": "flex",
          "zIndex": 98,
        }
      }
    >
      <div
        className="sc-bdvvtL gTyiTZ"
        color="#a89984"
        dir="right"
      />
      <div
        className="sc-gsDKAQ sc-dkPtRN ktNGwk einxge"
        dir="right"
      >
        r
      </div>
    </div>
  </div>
</div>
`;
